water/lib/Dialect/Wave/Transforms/LoweringPatterns.cpp

```diff
@@ -416,3 +419,147 @@ void wave::populateWaveMmaLoweringPatterns(WaveTypeConverter &typeConverter,
+
+    // Get the iterator symbol (e.g., "K").
+    wave::WaveSymbolAttr iteratorSymbol = op.getIterator();
+    StringRef symbolName = iteratorSymbol.getName();
+
+    // Find the tiling constraint for this iterator symbol.
+    func::FuncOp parentFunc = op->getParentOfType<func::FuncOp>();
+    if (!parentFunc) {
+      return rewriter.notifyMatchFailure(op, "iterate op not in function");
+    }
+
+    // Look for tiling constraints in function attributes.
+    ArrayAttr constraints = parentFunc->getAttrOfType<ArrayAttr>(
+        wave::WaveDialect::kWaveConstraintsAttrName);
+    if (!constraints) {
+      return rewriter.notifyMatchFailure(
+          op, "no wave constraints found in function");
+    }
+
+    // Get the dimension size (e.g., K = 640) from hyperparameters.
+    std::optional<SmallVector<int64_t>> resolvedDims =
+        wave::resolveSymbolNames(iteratorSymbol, hyperparams);
+    if (!resolvedDims || resolvedDims->size() != 1) {
+      return rewriter.notifyMatchFailure(
+          op, "iterator symbol not found in hyperparameters");
+    }
+    int64_t dimSize = resolvedDims->front();
+
+    // Find tiling constraint for this dimension to get tile_size.
+    std::optional<int64_t> tileSize;
+    for (Attribute constraintAttr : constraints) {
+      if (auto tilingConstraint =
+              dyn_cast<wave::TilingConstraintAttr>(constraintAttr)) {
+        wave::WaveSymbolAttr constraintDim = tilingConstraint.getDim();
+        if (constraintDim.getName() == symbolName) {
+          wave::WaveExprListAttr tileSizeAttr = tilingConstraint.getTileSize();
+          AffineMap tileSizeMap = tileSizeAttr.getMap();
+          ArrayRef<Attribute> tileSizeSymbols = tileSizeAttr.getSymbols();
+
+          // Evaluate the tile size using hyperparameters.
+          std::optional<SmallVector<int64_t>> evaluatedTileSize =
+              wave::evaluateMapWithHyperparams(tileSizeMap, tileSizeSymbols,
+                                               hyperparams);
+          if (!evaluatedTileSize) {
+            return rewriter.notifyMatchFailure(
+                op, "failed to evaluate tile size from tiling constraint");
+          }
+          if (evaluatedTileSize->size() != 1) {
+            return rewriter.notifyMatchFailure(
+                op, "tile size must be single value");
+          }
+          tileSize = (*evaluatedTileSize)[0];
+          break;
+        }
+      }
+    }
+
+    if (!tileSize) {
+      return rewriter.notifyMatchFailure(
+          op, "no tiling constraint found for iterator symbol");
+    }
+
+    // TODO(tyb): we reject non-exact division for now, which should require
+    // peeling or padding to be correct.
+    if (*tileSize == 0) {
+      return rewriter.notifyMatchFailure(op, "tile size cannot be zero");
+    }
+    if (dimSize % *tileSize != 0) {
+      return op.emitOpError("non-exact division not supported to prevent "
+                            "potential out-of-bounds access");
```

Comment:
Not for this patch, but can we think of a mechanism to make these better visible to the caller from python than an obscure message that only shows up in C++ debug build?